### PR TITLE
example: modal llm inference from dbt python model

### PR DESCRIPTION
**Pull request by kramstrom**
### Type of Change

- [x] New example
- [x] Example updates (Bug fixes, new features, etc.)
- [ ] Other (changes to the codebase, but not to examples)

## Checklist

- [ ] Example is testable in synthetic monitoring system, or `lambda-test: false` is added to example frontmatter (`---`)
  - [x] Example is tested by executing with `modal run` or an alternative `cmd` is provided in the example frontmatter (e.g. `cmd: ["modal", "deploy"]`)
  - [ ] Example is tested by running with no arguments or the `args` are provided in the example frontmatter (e.g. `args: ["--prompt", "Formula for room temperature superconductor:"]`
- [x] Example is documented with comments throughout, in a [_Literate Programming_](https://en.wikipedia.org/wiki/Literate_programming) style.
- [x] Example does _not_ require third-party dependencies to be installed locally
- [x] Example pins its dependencies
  - [x] Example pins container images to a stable tag, not a dynamic tag like `latest`
  - [ ] Example specifies a `python_version` for the base image, if it is used
  - [x] Example pins all dependencies to at least minor version, `~=x.y.z` or `==x.y`
  - [x] Example dependencies with `version < 1` are pinned to patch version, `==0.y.z`

## Outside contributors

You're great! Thanks for your contribution.

### CODE REVIEW

File: 06_gpu_and_ml/llm-serving/trtllm_llama.py [hunk 1]
```diff
@@ -567,7 +567,7 @@ def main():
 
 class GenerateRequest(pydantic.BaseModel):
     prompts: list[str]
-    settings: Optional[dict]
+    settings: Optional[dict] = None
```
Comment (kramstrom) [inline 06_gpu_and_ml/llm-serving/trtllm_llama.py:570]: Added `= None` as otherwise this wasn't really optional

Comment (charlesfrye) [reply 06_gpu_and_ml/llm-serving/trtllm_llama.py:570]: In Python typing, `Optional` means "can be `None`" and is distinct from "optional argument", i.e. argument with a default.

**Comment by charlesfrye** [comment]
looks like there's a few unrelated deprecation errors from client changes -- e.g. https://github.com/modal-labs/modal-examples/actions/runs/10014558567/job/27684478855?pr=816#step:4:314

File: 06_gpu_and_ml/llm-serving/trtllm_llama.py (hunk 1 above)
Comment (kramstrom) [reply 06_gpu_and_ml/llm-serving/trtllm_llama.py:570]: True, I guess we could change the `curl` example doc below instead to include a `settings = None` param if you prefer. With fastapi v2 either this has to be `None` or we need to pass it explicitly in the request, so one of the two places needs an update

**Comment by kning** [comment]
very cool! i also have this dlt modal example running on our data currently: https://github.com/modal-labs/modal-etl/blob/main/dlt/postgres_pipeline.py. i think could be a great blog post to combine those two into one e2e etl example.

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
+# ---
+# cmd: ["modal", "run", "10_integrations/dbt_modal_inference/duck_db_modal_inference.py"]
```
Comment (charlesfrye) [inline 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:2]: this shouldn't be necessary, right? -- this should be the default command for anything testable

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+# ---
+# # dbt python model with llm inference
+#
+# In this example we demonstrate how you could combine dbt python models with llm
```
Comment (charlesfrye) [inline 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:6]: I think it's a bit confusing to use "models" multiple ways here. Are there alternative ways to describe what DBT is doing? Is it really just providing data models?

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+#
+# ## Overview
+#
+# This example runs [dbt](https://docs.getdbt.com/docs/introduction) with a [duckdb](https://duckdb.org)
+# backend directly on top of modal, but could be translated to run on any dbt-compatible
+# database that supports python models. Similarly you could make these requests from UDFs
+# directly in SQL instead if you don't want to use python models.
+#
+# In this example we use an llm deployed in a previous example: [Serverless TensorRT-LLM (LLaMA 3 8B)](https://modal.com/docs/examples/trtllm_llama)
+# but you could easily swap this for whichever modal function you wish. We use this to classify the sentiment
```
Comment (charlesfrye) [inline 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:17]: nit: we generally capitalize Modal and Function -- to separate the proper noun/product component "Function" from the normal old functions in Python

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+# for free-text product reviews and aggregate them in subsequent dbt sql models.
```
Comment (charlesfrye) [inline 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:18]: Did you also generate the reviews with the LLM? That's a neat point to include and justifies the use of a larger model.

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+#
+#
+# ## dbt & image configuration
+#
+# We set up the environment variables necessary for dbt and
+# create a slim debian and install the packages necessary to run.
```
Comment (charlesfrye) [inline 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:24]: See the [updated description](https://github.com/modal-labs/modal-examples/commit/da92247c52fd5a6e32f9cbad8567704ca733a950) of a similar section of the other DBT example for a sample of how we prefer to explain the environment setup.

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+# └──────────────────────────────────────────┴──────┴───────────────────────┴─────────┘
+# ```
+#
+# ## Python dbt model
+# The python dbt model in `dbt_modal_inference_proj/models/product_reviews_sentiment.py` is quite simple.
+# It configures dbt to store the results locally in parquet, fetches the inference url,
+# uses a batch reader to iterate over the inputs and call the `batcher` method
+# and adds a new field from the LLM
+#
+# ```python
```
Comment (charlesfrye) [inline 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:200]: Altogether, there's quite a bit of inlined Python here and we would need to duplicate any changes. Do we need the code here or can we point readers to the files, where we can put comments?

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py (hunk 3 above)
Comment (kramstrom) [reply 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:6]: I agree it's a bit confusing, but calling them sql / python / data models really is the dbt way. I can see if I can make it a bit clearer though

File: 10_integrations/dbt_modal_inference/duck_db_modal_inference.py (hunk 7 above)
Comment (kramstrom) [reply 10_integrations/dbt_modal_inference/duck_db_modal_inference.py:200]: That would work, do you think we should link to the files on github directly?

**Pull request merged**